src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeClientStream.Windows.cs

```diff
@@ -79,14 +55,50 @@ private bool TryConnect(int timeout, CancellationToken cancellationToken)
             {
                 int errorCode = Marshal.GetLastWin32Error();
 
-                // Handle the possible race condition of someone else connecting to the server 
-                // between our calls to WaitNamedPipe & CreateFile.
-                if (errorCode == Interop.Errors.ERROR_PIPE_BUSY)
+                if (errorCode == Interop.Errors.ERROR_FILE_NOT_FOUND ||
+                    errorCode == Interop.Errors.ERROR_PIPE_BUSY)
                 {
-                    return false;
+                    // ERROR_PIPE_BUSY
+                    // Handle the possible race condition of someone else connecting to the server 
+                    // between our calls to WaitNamedPipe & CreateFile.
+                    
+                    // ERROR_FILE_NOT_FOUND
+                    // Server is not yet created
+
+                    // Fallback to WaitNamedPipe
+                    if (!Interop.Kernel32.WaitNamedPipe(_normalizedPipePath, timeout))
+                    {
+                        int waitErrorCode = Marshal.GetLastWin32Error();
+
+                        // Server is not yet created
+                        if (waitErrorCode == Interop.Errors.ERROR_FILE_NOT_FOUND)
+                        {
+                            return false;
+                        }
+
+                        // The timeout has expired.
+                        if (waitErrorCode == Interop.Errors.ERROR_SUCCESS)
+                        {
+                            if (cancellationToken.CanBeCanceled)
+                            {
+                                // It may not be real timeout.
+                                return false;
+                            }
+                            throw new TimeoutException();
+                        }
+
+                        throw Win32Marshal.GetExceptionForWin32Error(waitErrorCode);
+                    }
+                    else
+                    {
+                        // We have not timeouted yet, but we need to update the timeout value.
+                        return false;
```

Comment:
We should be trying to open the pipe as it should be available in this case. If we exit out we might not actually open the pipe when we would have before.

Reply:
Yes, while TryConnect is called in a loop if it return false the code bails out of the loop.

---
In reply to: [144418419](https://github.com/dotnet/corefx/pull/24616#discussion_r144418419) [](ancestors = 144418419)

Reply:
It only bails out when it returns true, it does not bail out on false.